sims/benchmark-simuPOP.py

```diff
@@ -0,0 +1,330 @@
+
+logfile.write("Locus positions:\n")
+logfile.write(str(locus_position)+"\n")
+logfile.write("----------\n")
+logfile.flush()
+
+###
+# modified from http://simupop.sourceforge.net/manual_svn/build/userGuide_ch5_sec9.html
+
+class GammaDistributedFitness:
+    def __init__(self, shape, scale):
+        # mean is shape/scale
+        self.coefMap = {}
+        # these are called alpha/beta by `random.gammavariate` but are
+        # implemented as shape/scale paramterization
+        #
+        self.alpha = shape
+        self.beta = scale
+    def __call__(self, loc, alleles):
+        # because s is assigned for each locus, we need to make sure the
+        # same s is used for fitness of genotypes 01 (1-s) and 11 (1-2s)
+        # at each locus
+        if loc in self.coefMap:
+            s = self.coefMap[loc]
+        else:
+            s = random.gammavariate(self.alpha, self.beta)
+            self.coefMap[loc] = s
+        # print(str(loc)+":"+str(alleles)+"\n")
+        # needn't return fitness for alleles=(0,0) as simupop knows that's 1
+        if 0 in alleles:
+            return max(0.0, 1. - s)
+        else:
+            return max(0.0, 1. - 2.*s)
+
+time_dict = {'prepping': None,
+             'simulating': None,
+             'finalizing': None}
+
+start = time.time()
+
+init_geno=[sim.InitGenotype(freq=init_freqs[k],loci=init_classes[k])
+           for k in range(len(init_freqs))]
+
+pop = sim.Population(
+        size=[args.popsize]*npops, 
+        loci=[args.nloci], 
+        lociPos=locus_position,
+        infoFields=['ind_id','fitness','migrate_to'])
+
+id_tagger = sim.IdTagger()
+id_tagger.apply(pop)
+
+# set up recomb collector
+first_gen = pop.indInfo("ind_id")
+init_ts = msprime.simulate(2*len(first_gen),
+                           length=max(locus_position))
+haploid_labels = [(k,p) for k in first_gen
+                        for p in (0,1)]
+node_ids = {x:j for x, j in zip(haploid_labels, init_ts.samples())}
+rc = RecombCollector(ts=init_ts, node_ids=node_ids,
+                     locus_position=locus_position)
+end_prep = time.time()
+time_dict['prepping'] = end_prep - start
+
+pop.evolve(
+    initOps=[
+        sim.InitSex(),
+    ]+init_geno,
+    preOps=[
+        sim.PyOperator(lambda pop: rc.increment_time() or True),
+        sim.SNPMutator(u=args.sel_mut_rate, v=args.sel_mut_rate),
+        # so that selector returns, for f_i fitness values, \prod_i f_i
+        sim.PyMlSelector(GammaDistributedFitness(shape=args.gamma_shape,
+                                                 scale=args.gamma_scale),
+                         mode=sim.MULTIPLICATIVE),
+    ],
+    matingScheme=sim.RandomMating(
+        ops=[
+            id_tagger,
+            sim.Recombinator(intensity=args.recomb_rate,
+                output=rc.collect_recombs,
+                infoFields="ind_id"),
+        ] ),
+    postOps=[
+        sim.Stat(numOfSegSites=sim.ALL_AVAIL, step=50),
```

Comment:
change to calc segregrating sites and fixations (as in KT)

up step size or possibly just do at end